### DISCUSSION TITLE
"Cannot assign requested address" using localhost:<PORT> inside container

### DISCUSSION THREAD

**Discussion by esabol** [community/community#26331] [Actions]
<p>This is driving me crazy. I’ve trying to migrate from Travis CI to GitHub Actions CI for months, but I cannot get our test suite to pass when building and testing inside of a container. If I don’t use a container and use use “os: ubuntu-”, the tests pass. If I use a container, however, I get the dreaded “Cannot assign requested address” error. My googling brought up this issue:</p>
<aside class="onebox githubissue" data-onebox-src="https://github.com/actions/virtual-environments/issues/3185">
  <header class="source">

      <a href="https://github.com/actions/virtual-environments/issues/3185" target="_blank" rel="noopener nofollow ugc">github.com/actions/virtual-environments</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/actions/virtual-environments/issues/3185" target="_blank" rel="noopener nofollow ugc">Reverse name lookup is broken for current hostname in ubuntu VMs</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2021-04-14" data-time="15:59:36" data-timezone="UTC">03:59PM - 14 Apr 21 UTC</span>
      </div>

        <div class="date">
          closed <span class="discourse-local-date" data-format="ll" data-date="2021-04-26" data-time="11:55:20" data-timezone="UTC">11:55AM - 26 Apr 21 UTC</span>
        </div>

      <div class="user">
        <a href="https://github.com/lhotari" target="_blank" rel="noopener nofollow ugc">
          <img alt="lhotari" src="https://user-images.githubusercontent.com/22986767/181096978-cf8dae92-ea05-43da-90fb-909a5c6f8795.jpeg" class="onebox-avatar-inline" width="20" height="20">
          lhotari
        </a>
      </div>
    </div>

    <div class="labels">
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">
          awaiting-deployment
        </span>
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">
          OS: Ubuntu
        </span>
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">
          Area: Common Tools
        </span>
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">
          investigate
        </span>
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">
          external
        </span>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">**Description**  

Reverse name lookup is broken for the current hostname. Thi<span class="show-more-container"><a href="" rel="noopener" class="show-more">…</a></span><span class="excerpt hidden">s leads to odd issues. Here are some reports from GitHub Community:
* https://github.community/t/github-action-and-oserror-errno-99-cannot-assign-requested-address/173973
* https://github.community/t/volatile-local-ip-addresses-during-action-run/174362
* https://github.community/t/github-actions-timeout/174533

This is a problem report in apache/pulsar: https://github.com/apache/pulsar/issues/10232
and workaround in apache/pulsar https://github.com/apache/pulsar/pull/10233

I have suggested this general workaround to the problem:
```
- name: Add the current IP address, long hostname and short hostname record to /etc/hosts file
  run: |
     echo -e "$(ip addr show eth0 | grep "inet\b" | awk '{print $2}' | cut -d/ -f1)\t$(hostname -f) $(hostname -s)" | sudo tee -a /etc/hosts
```

**Question, Bug, or Feature?**:  Bug

**Virtual environments affected**  
- [x] Ubuntu 16.04
- [x] Ubuntu 18.04
- [x] Ubuntu 20.04
- [ ] macOS 10.15
- [ ] macOS 11.0
- [ ] Windows Server 2016 R2
- [ ] Windows Server 2019

**Expected behavior**  

`hostname -i`, `nslookup $(hostname -s)` and `nslookup $(hostname -l)`  should return the IP address of the current hostname.

**Actual behavior**  

`hostname -i`, `nslookup $(hostname -s)` or `nslookup $(hostname -l)` returns an IP address that changes and the IP address doesn't belong to the current host.

**Repro steps**  

Add this command to some action and run it `ip a; while [ 1 ]; do nslookup $(hostname); sleep 3; done`

Example output
```
1: lo: &lt;LOOPBACK,UP,LOWER_UP&gt; mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: &lt;BROADCAST,MULTICAST,UP,LOWER_UP&gt; mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:0d:3a:11:ff:4c brd ff:ff:ff:ff:ff:ff
    inet 10.1.1.35/16 brd 10.1.255.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::20d:3aff:fe11:ff4c/64 scope link 
       valid_lft forever preferred_lft forever
3: docker0: &lt;NO-CARRIER,BROADCAST,MULTICAST,UP&gt; mtu 1500 qdisc noqueue state DOWN group default 
    link/ether 02:42:b1:57:12:f2 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever

Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
Name:   fv-az132-427.w5jmtgevnbgefpdoz44oxoejwc.bx.internal.cloudapp.net
Address: 10.1.1.83

Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
Name:   fv-az132-427.w5jmtgevnbgefpdoz44oxoejwc.bx.internal.cloudapp.net
Address: 10.1.1.83

Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
Name:   fv-az132-427.w5jmtgevnbgefpdoz44oxoejwc.bx.internal.cloudapp.net
Address: 10.1.1.83

Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
Name:   fv-az132-427.w5jmtgevnbgefpdoz44oxoejwc.bx.internal.cloudapp.net
Address: 10.1.1.83

Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
Name:   fv-az132-427.w5jmtgevnbgefpdoz44oxoejwc.bx.internal.cloudapp.net
Address: 10.1.0.37

Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
Name:   fv-az132-427.w5jmtgevnbgefpdoz44oxoejwc.bx.internal.cloudapp.net
Address: 10.1.0.37

Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
Name:   fv-az132-427.w5jmtgevnbgefpdoz44oxoejwc.bx.internal.cloudapp.net
Address: 10.1.0.37

Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
Name:   fv-az132-427.w5jmtgevnbgefpdoz44oxoejwc.bx.internal.cloudapp.net
Address: 10.1.0.37
```

In the above case, the reverse lookup returns addresses 10.1.1.83 and 10.1.0.37 which aren't the ip addresses of the current host. The address of eth0 is 10.1.1.35 .</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>But that was supposedly fixed and resolved last year. I thought maybe ubuntu:14.04 wasn’t fixed (since it’s ancient and wasn’t listed in the issue), so I tried the workaround that folks used before it was fixed, but I can’t seem to get it working.</p>
<p>Can anyone help me, please?</p>
<p>Here’s a failing run:</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/gearman/gearmand/runs/6462158831">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/gearman/gearmand/runs/6462158831" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/22986767/181096979-f6ed441b-f91a-453b-a982-8b49dacb46e4.png" class="thumbnail onebox-avatar" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png">

<h3><a href="https://github.com/gearman/gearmand/runs/6462158831" target="_blank" rel="noopener nofollow ugc">Build software better, together</a></h3>

  <p>GitHub is where people build software. More than 83 million people use GitHub to discover, fork, and contribute to over 200 million projects.</p>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Here’s my workflow yml:</p>
<aside class="onebox githubblob" data-onebox-src="https://github.com/gearman/gearmand/blob/e42d60e587d60ac396461a75d6f66d4744e3a565/.github/workflows/c-cpp.yml">
  <header class="source">

      <a href="https://github.com/gearman/gearmand/blob/e42d60e587d60ac396461a75d6f66d4744e3a565/.github/workflows/c-cpp.yml" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/gearman/gearmand/blob/e42d60e587d60ac396461a75d6f66d4744e3a565/.github/workflows/c-cpp.yml" target="_blank" rel="noopener nofollow ugc">gearman/gearmand/blob/e42d60e587d60ac396461a75d6f66d4744e3a565/.github/workflows/c-cpp.yml</a></h4>

      <pre><code class="lang-yml">name: GitHub Actions CI

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    name: ${{ matrix.config.name }}
    runs-on: ${{ matrix.config.os }}
    container: ${{ matrix.config.container }}

    strategy:
      fail-fast: false
      matrix:
        config:
          - {name: 'ubuntu-14.04 gcc-4.8', os: ubuntu-latest, container: 'ubuntu:14.04', cc: 'gcc', cxx: 'g++', tag: '4.8', toolchain: 'ppa:ubuntu-toolchain-r/test'}
</code></pre>

  This file has been truncated. <a href="https://github.com/gearman/gearmand/blob/e42d60e587d60ac396461a75d6f66d4744e3a565/.github/workflows/c-cpp.yml" target="_blank" rel="noopener nofollow ugc">show original</a>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Thanks,<br>
Ed</p>

**Comment by esabol** [comment]
<p>Trying again… Can anyone help, please???</p>

**Comment by airtower-luna** [comment]
<p>Does the container image you’re using define an IPv6 address for <code>localhost</code> in <code>/etc/hosts</code>? The Docker containers on the runners don’t get IPv6, even on the loopback device.</p>

**Reply by nickeldan** [reply]
Is there no way to configure that?  That is, is there any way to configure the workflow get an IPv6 loopback address?

**Reply by airtower-luna** [reply]
Unfortunately not. :confused: Without the job container you could try to reconfigure the Docker daemon enable IPv6 for the Docker networks at least, but I have no idea if that'd bump into any other issues.

See also: https://github.com/actions/runner-images/issues/668 It's a slightly different topic but closely related, note the "wontfix" tag.

**Comment by esabol** [comment]
<p>I think so? I “cat /etc/hosts” and run some other network commands in the workflow. Here’s the output:</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/gearman/gearmand/runs/6462158831#step:4:24">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/gearman/gearmand/runs/6462158831#step:4:24" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/22986767/181096984-84c21b87-84fc-4003-b61f-6696a8d099c9.png" class="thumbnail onebox-avatar" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png">

<h3><a href="https://github.com/gearman/gearmand/runs/6462158831#step:4:24" target="_blank" rel="noopener nofollow ugc">Build software better, together</a></h3>

  <p>GitHub is where people build software. More than 83 million people use GitHub to discover, fork, and contribute to over 200 million projects.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<h1>
<a name="h-127001-localhost-1-localhost-ip6-localhost-ip6-loopback-fe000-ip6-localnet-ff000-ip6-mcastprefix-ff021-ip6-allnodes-ff022-ip6-allrouters-1721802-e1755e97477e-1" class="anchor" href="#h-127001-localhost-1-localhost-ip6-localhost-ip6-loopback-fe000-ip6-localnet-ff000-ip6-mcastprefix-ff021-ip6-allnodes-ff022-ip6-allrouters-1721802-e1755e97477e-1"></a>127.0.0.1 localhost<br>
::1 localhost ip6-localhost ip6-loopback<br>
fe00::0 ip6-localnet<br>
ff00::0 ip6-mcastprefix<br>
ff02::1 ip6-allnodes<br>
ff02::2 ip6-allrouters<br>
172.18.0.2 e1755e97477e</h1>
<h1>
<a name="h-1-lo-loopbackuplower_up-mtu-65536-qdisc-noqueue-state-unknown-group-default-qlen-1000-linkloopback-000000000000-brd-000000000000-inet-1270018-scope-host-lo-valid_lft-forever-preferred_lft-forever-5-eth0if6-broadcastmulticastuplower_up-mtu-1500-qdisc-noqueue-state-up-group-default-linkether-0242ac120002-brd-ffffffffffff-inet-172180216-brd-17218255255-scope-global-eth0-valid_lft-forever-preferred_lft-forever-2" class="anchor" href="#h-1-lo-loopbackuplower_up-mtu-65536-qdisc-noqueue-state-unknown-group-default-qlen-1000-linkloopback-000000000000-brd-000000000000-inet-1270018-scope-host-lo-valid_lft-forever-preferred_lft-forever-5-eth0if6-broadcastmulticastuplower_up-mtu-1500-qdisc-noqueue-state-up-group-default-linkether-0242ac120002-brd-ffffffffffff-inet-172180216-brd-17218255255-scope-global-eth0-valid_lft-forever-preferred_lft-forever-2"></a>1: lo: &lt;LOOPBACK,UP,LOWER_UP&gt; mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000<br>
link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00<br>
inet 127.0.0.1/8 scope host lo<br>
valid_lft forever preferred_lft forever<br>
5: eth0@if6: &lt;BROADCAST,MULTICAST,UP,LOWER_UP&gt; mtu 1500 qdisc noqueue state UP group default<br>
link/ether 02:42:ac:12:00:02 brd ff:ff:ff:ff:ff:ff<br>
inet 172.18.0.2/16 brd 172.18.255.255 scope global eth0<br>
valid_lft forever preferred_lft forever</h1>
<p>‘hostname -i’ shows ‘172.18.0.2’<br>
‘hostname -I’ shows '172.18.0.2 ’</p>

**Comment by airtower-luna** [comment]
<p>My points is, if your software is trying to bind to the IPv6 address for <code>localhost</code> it can’t work in a container with no IPv6. Might be a thing to investigate.</p>

**Comment by esabol** [comment]
<p>I’m not sure how to investigate it. It’s trying to bind to “localhost”, certainly. What do I need to change in /etc/hosts to get it to work?</p>

**Comment by airtower-luna** [comment]
<p>Look at the code: Does it use IPv6 or IPv4/IPv6 dualstack? If yes, you could give it an IPv4 only mode.</p>
<p>For a very quick test you could remove the <code>localhost</code> name for <code>::1</code> with a command like this in your workflow:</p>
<pre><code class="lang-bash">sed -i -e 's/^::1\slocalhost\s\(.*\)/::1 \1/' /etc/hosts
</code></pre>

**Comment by esabol** [comment]
<p>I got a “Device or file busy” error using “sed -i”, so I changed it to this:</p>
<pre><code class="lang-nohighlight">        sed 's/^::1\s\s*localhost\s\(.*\)/::1 \1/' /etc/hosts &gt; /tmp/hosts.temp
        cp /tmp/hosts.temp /etc/hosts
        rm /tmp/hosts.temp
</code></pre>
<p>Here’s the new contents of <code>/etc/hosts</code> inside the container:</p>
<pre><code class="lang-nohighlight">127.0.0.1	localhost
::1 ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.18.0.2	7076974732ab
</code></pre>
<p>And here’s the output of <code>ip addr</code>:</p>
<pre><code class="lang-nohighlight">1: lo: &lt;LOOPBACK,UP,LOWER_UP&gt; mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
5: eth0@if6: &lt;BROADCAST,MULTICAST,UP,LOWER_UP&gt; mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:12:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.2/16 brd 172.18.255.255 scope global eth0
       valid_lft forever preferred_lft forever
</code></pre>
<p>I really had high hopes this would work, but it didn’t. :frowning: However, the error has now changed from “Cannot assign requested address” to “Connection refused”! So progress, I think! Any other suggestions?</p>
<pre><code class="lang-nohighlight">libtest/client.cc:273: in start() pid(23275) localhost:59298 ping(libtest/client.cc:273: Connection refused), additionally pid: 23315 is alive: true waited: 17 server started. exec: /__w/gearmand/gearmand/libtool --mode=execute /__w/gearmand/gearmand/./gearmand/gearmand --verbose=INFO --log-file=var/log/gearmand.log8ch4wj --pid-file=var/run/gearmand.pid2itON2 --port=59298 --exceptions -u root --listen=localhost stderr:
</code></pre>
<p>Link: <a href="https://github.com/gearman/gearmand/runs/6617520499?check_suite_focus=true#step:12:93" rel="noopener nofollow ugc">92</a></p>
<p>I don’t get why the same code and the same commands work just fine in Travis CI and in the Docker engine running on my Linux box, but I appreciate the assistance!!!</p>

**Comment by esabol** [comment, marked as answer]
<p>Never mind! I just got it to work. Finally!</p>
<p>For some reason our test suite code to find an available port isn’t working inside the container, but if I force it to use a specific port number that I know is available, then it works! All tests pass! Yippee!</p>
